### PR TITLE
Auto-convert attributes to/from python types

### PR DESCRIPTION
**Pull request by t-wissmann**
In the python bindings, automatically convert the attribute values from
(resp. to) their corresponding python types when writing (resp. reading)
attributes.

This requires touching (but fortunately simplifying) a lot of assertions
in the test cases. Now, working with the contents of hlwm.attr should
feel as if one had the hlwm attributes available as python variables.

A few new test cases at the end of test_python_binds.py illustrate the
usage for all types that are available at the moment.

### CODE REVIEW

File: python/herbstluftwm/types.py [hunk 1]
```diff
@@ -48,3 +48,81 @@ def adjusted(self, dx=0, dy=0, dw=0, dh=0):
...
+
+        HlwmType('rectangle',
+                 Rectangle.from_user_str,
+                 Rectangle.to_user_str,
+                 lambda x: isinstance(x, Rectangle)),
+
+        HlwmType('string',
+                 lambda x: x,
+                 lambda x: x,
+                 lambda x: isinstance(x, str)),
```
Comment (The-Compiler) [inline python/herbstluftwm/types.py:115]: Personally I find this a bit hard to follow, because the positional lambdas are different "methods", and then there's things scattered in different places in this file (e.g. the `bool_from_user_str` below).

Have you considered making those subclasses of `HlwmType` (with the name being a constant class attribute, and the lambdas being methods instead)?

File: tests/test_layout.py [hunk 2]
```diff
@@ -369,23 +369,23 @@ def test_cycle_stays_in_floating_layer(hlwm):
     for winid in floating:
         assert hlwm.attr.clients.focus.winid() == winid
         # the floating layer always stays focused
-        assert hlwm.attr.tags.focus.floating_focused() == hlwm.bool(True)
+        assert hlwm.attr.tags.focus.floating_focused() is True
```
Comment (The-Compiler) [inline tests/test_layout.py:372]:
```suggestion
        assert hlwm.attr.tags.focus.floating_focused()
```

maybe? We're not really testing the `floating_focused` getter type here, and I assume in C(++) you'd also write `if (floating_focused)` rather than `if (floating_focused == true)`.

(Same for all other instances of `is True` and `is False`, except maybe `test_implicit_type_conversion_bool` because we're really interested in the type there)

File: tests/test_layout.py [hunk 3]
```diff
@@ -369,23 +369,23 @@ def test_cycle_stays_in_floating_layer(hlwm):
...
         hlwm.call(['cycle', '+1'])
 
     # we're back a the first floating client:
     assert hlwm.attr.clients.focus.winid() == floating[0]
-    assert hlwm.attr.tags.focus.floating_focused() == hlwm.bool(True)
+    assert hlwm.attr.tags.focus.floating_focused() is True
```
Comment (The-Compiler) [inline tests/test_layout.py:377]: ditto

File: tests/test_monitor.py [hunk 4]
```diff
@@ -158,7 +158,7 @@ def test_move_monitor(hlwm):
 
 
 def test_move_monitor_completion(hlwm):
-    assert hlwm.attr.monitors.focus.geometry() in hlwm.complete('move_monitor 0')
+    assert hlwm.get_attr('monitors.focus.geometry') in hlwm.complete('move_monitor 0')
```
Comment (The-Compiler) [inline tests/test_monitor.py:161]: What's the difference?

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Lovely! I added a couple of comments.

File: python/herbstluftwm/types.py (hunk 1 above)
Comment (t-wissmann) [reply python/herbstluftwm/types.py]: I considered this, but it felt odd to have subclasses with `@staticmethod`s only (and undefined `@staticmethod`s in `HlwmType`). Also I had the impression that it's more write effort.

I agree  that the positional arguments are hard to read. I have just replaced them with named arguments; is this more readable?

File: tests/test_layout.py (hunk 2 above)
Comment (t-wissmann) [reply tests/test_layout.py:372]: Hm you're probably right, but it feels fragile because the type is detected at run-time and returning the string `"false"` here would make the `assert` hold. (Also, I think `assert … is True` reads quite well)

File: tests/test_monitor.py (hunk 4 above)
Comment (t-wissmann) [reply tests/test_monitor.py]: the completions is always a list of strings, but `hlwm.attr.monitors.focus.geometry()` is now of type `Rectangle`. I've just replaced it with `str(hlwm.attr.monitors.focus.geometry)`, maybe this emphasizes more what's going on.

**Comment by t-wissmann** [comment]
Thanks for reviewing the PR :-) It's good to have you python expertise :)

File: python/herbstluftwm/types.py (hunk 1 above)
Comment (The-Compiler) [reply python/herbstluftwm/types.py]: Definitely better! I'd still prefer classes (then you could also e.g. have proper type annotations instead of describing the types in the comments), but no strong opinions either way.

**Review by blarz — CHANGES_REQUESTED** [review]
Small typo, other than that, I'd leave it to @The-Compiler  to approve this change

File: python/herbstluftwm/types.py [hunk 5]
```diff
@@ -48,3 +48,87 @@ def adjusted(self, dx=0, dy=0, dw=0, dh=0):
...
+                 from_user_str=lambda x: x,
+                 to_user_str=lambda x: x,
+                 is_instance=lambda x: isinstance(x, str)),
+    ]
+
+    return types
+
+
+def bool_from_user_str(bool_string):
+    """Parse a the string description of a hlwm boolean to
```
[the remark this answers is not in the document]
Comment (blarz) [reply python/herbstluftwm/types.py]: Thanks!

**Review by blarz — COMMENTED** [review]
Removing my request for changes :)

**Review by The-Compiler — APPROVED** [review]
Looks good, sorry for the delay!
